sympy/tensor/array/array_comprehension.py

```diff
@@ -272,4 +273,59 @@ def _array_subs(arr, var, val):
                     list_gen.append(list_expr.subs(var, val))
                 else:
                     list_gen.append(_array_subs(list_expr, var, val))
-        return ImmutableDenseNDimArray(list_gen)
+        return list_gen
+
+    def tolist(self):
+        """Transform the expanded array to a list
+
+        Raises
+        ======
+
+        ValueError : When there is a symbolic dimension
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy import symbols
+        >>> i, j = symbols('i j')
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> a.tolist()
+        [[11, 12, 13], [21, 22, 23], [31, 32, 33], [41, 42, 43]]
+        """
+        if self.is_numeric:
+            return self._expand_array()
+
+        raise ValueError("A symbolic array cannot be expanded to a list")
+
+    def tomatrix(self):
+        """Transform the expanded array to a matrix
+
+        Raises
+        ======
+
+        ValueError : When there is a symbolic dimension
+        ValueError : When the rank of the expanded array is not equal to 2
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy import symbols
+        >>> i, j = symbols('i j')
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> a.tomatrix()
+        Matrix([
+        [11, 12, 13],
+        [21, 22, 23],
+        [31, 32, 33],
+        [41, 42, 43]])
+        """
+        from sympy.matrices import Matrix
+
+        if not self.is_numeric:
+            raise ValueError("A symbolic array cannot be expanded to a matrix")
+        if self.rank() != 2:
```

Comment:
I think there was something like, `self._rank`, can it be used here directly? As far as I remember `rank` also calls `self._rank`. A short explanation would help if the said cannot be done. Thanks.

Reply:
Yes, There is nearly no difference between `self._rank` and `self.rank()`. BUt it would be better to keep the same coding convention. So that is a mistake. Thank for pointing it out!

Reply:
OK